product-accelerators/apim/components/org.wso2.healthcare.apim.backendauth/src/main/java/org/wso2/healthcare/apim/backendauth/PrivateKeyJWTBackendAuthenticator.java

```diff
@@ -0,0 +1,313 @@
+import java.io.UnsupportedEncodingException;
+import java.net.HttpURLConnection;
+import java.security.Key;
+import java.security.KeyStore;
+import java.security.cert.Certificate;
+import java.security.cert.X509Certificate;
+import java.security.interfaces.RSAPrivateKey;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.Map;
+import java.util.UUID;
+
+/**
+ * Mediator to authenticate with the backend using private key JWT.
+ */
+public class PrivateKeyJWTBackendAuthenticator extends BackendAuthenticator {
+
+    private static final Log log = LogFactory.getLog(PrivateKeyJWTBackendAuthenticator.class);
+    private static final JsonParser parser = new JsonParser();
+
+    public PrivateKeyJWTBackendAuthenticator() throws OpenHealthcareException {
+    }
+
+    public boolean mediate(MessageContext messageContext) {
+
+        log.info("PrivateKeyJWTBackendAuthenticator mediator is started.");
+
+        String accessToken;
+        String tokenEndpoint = backendAuthConfig.getAuthEndpoint();
+        String keyAlias = backendAuthConfig.getPrivateKeyAlias();
+        char[] keyStorePass = CarbonUtils.getServerConfiguration().getFirstProperty(
+                "Security.KeyStore.Password").toCharArray();
+        char[] trustStorePass = CarbonUtils.getServerConfiguration().getFirstProperty(
+                "Security.TrustStore.Password").toCharArray();
+        String clientId = backendAuthConfig.getClientId();
+        int tenantId = Integer.parseInt(messageContext.getProperty(Constants.TENANT_INFO_ID).toString());
+        KeyStoreManager keyStoreManager = KeyStoreManager.getInstance(tenantId);
+        Key privateKey;
+        Certificate publicCert;
+
+        if (log.isDebugEnabled()) {
+            log.debug("Configured client ID: " + clientId);
+        }
+
+        try {
+            KeyStore primarykeyStore = keyStoreManager.getPrimaryKeyStore();
+            privateKey = primarykeyStore.getKey(keyAlias, keyStorePass);
+            publicCert = keyStoreManager.getDefaultPrimaryCertificate();
+        } catch (Exception e) {
+            log.error("Error occurred while retrieving private key from keystore.", e);
+            throw new OpenHealthcareRuntimeException(e);
+        }
+
+        Token token = TokenManager.getToken(clientId, tokenEndpoint);
+        if (token == null || !token.isActive()) {
+            if (log.isDebugEnabled()) {
+                log.debug("Access token not available in TokenManager.");
+            }
+            try {
+                token = getAndAddNewToken(clientId, tokenEndpoint, messageContext, privateKey, publicCert, trustStorePass);
+            } catch (OpenHealthcareException e) {
+                log.error("Error occurred while retrieving access token.",e);
+                throw new OpenHealthcareRuntimeException(e);
+            }
+        }
+        accessToken = token.getAccessToken();
+
+        messageContext.setProperty(Constants.PROPERTY_ACCESS_TOKEN, accessToken);
```

Comment:
Why setting as a property?

Reply:
removed in [b8da50d](https://github.com/wso2/healthcare-accelerator/pull/7/commits/b8da50dbe832d38ee8259489c5c03798ba7ca81d)